File: src/ranker.ts
import type { TriageFeatures } from './triage.js';
import type { FloorReservation } from './rules.js';
import type { JournalEntry } from './journal.js';

/** Normalised feature vector for scoring. All values in [0, 1]. */
export interface FeatureVector {
  deadline_proximity: number;
  has_amount: number;
  waiting_on_user: number;
  urgency: number;
}

const FEATURE_KEYS: (keyof FeatureVector)[] = ['deadline_proximity', 'has_amount', 'waiting_on_user', 'urgency'];

/** Per-feature weights used by the scorer. */
export type FeatureWeights = Record<keyof FeatureVector, number>;

/** Min/max bounds for weight clamping. */
export const WEIGHT_BOUNDS: Record<keyof FeatureVector, { min: number; max: number }> = {
  deadline_proximity: { min: 0.1, max: 3.0 },
  has_amount: { min: 0.1, max: 3.0 },
  waiting_on_user: { min: 0.1, max: 3.0 },
  urgency: { min: 0.2, max: 3.0 },
};

export const DEFAULT_WEIGHTS: FeatureWeights = {
  deadline_proximity: 1.0,
  has_amount: 0.5,
  waiting_on_user: 0.8,
  urgency: 1.0,
};

/** Extract a normalised feature vector from triage features. */
export function extractFeatureVector(features: TriageFeatures, now: Date = new Date()): FeatureVector {
  // Deadline proximity: 1.0 if <= 24h, linear decay to 0 at 168h (7 days), 0 if no deadline
  let deadline_proximity = 0;
  if (features.deadline) {
    const deadline = new Date(features.deadline);
    const hoursUntil = (deadline.getTime() - now.getTime()) / (1000 * 60 * 60);
    if (hoursUntil >= 0 && hoursUntil <= 168) {
      deadline_proximity = Math.max(0, 1 - hoursUntil / 168);
    }
  }

  const urgencyMap: Record<string, number> = { high: 1.0, medium: 0.5, low: 0.0 };

  return {
    deadline_proximity,
    has_amount: features.amount != null ? 1 : 0,
    waiting_on_user: features.waiting_on_user ? 1 : 0,
    urgency: urgencyMap[features.urgency] ?? 0,
  };
}

/** Compute a weighted score for a feature vector. Higher = more important. */
export function scoreCandidate(fv: FeatureVector, weights: FeatureWeights): number {
  let total = 0;
  for (const key of FEATURE_KEYS) {
    total += fv[key] * weights[key];
  }
  return total;
}

/** Per-feature score breakdown. */
export interface ScoreBreakdown {
  deadline_proximity: number;
  has_amount: number;
  waiting_on_user: number;
  urgency: number;
  total: number;
}

function computeBreakdown(fv: FeatureVector, weights: FeatureWeights): ScoreBreakdown {
  const breakdown: Record<string, number> = {};
  let total = 0;
  for (const key of FEATURE_KEYS) {
    const contribution = fv[key] * weights[key];
    breakdown[key] = contribution;
    total += contribution;
  }
  return { ...breakdown, total } as ScoreBreakdown;
}

/**
 * Learn weights from swipe history in journal entries.
 * Approvals increase weights for features that were present;
 * rejections/defers decrease them. Weights are clamped.
 */
export function learnWeights(entries: JournalEntry[]): FeatureWeights {
  const weights = { ...DEFAULT_WEIGHTS };

  // Filter to decision entries that have features
  const decisions = entries.filter(
    (e) => e.kind === 'decision' && (e.decision === 'approve' || e.decision === 'reject' || e.decision === 'defer') && e.features,
  );

  if (decisions.length === 0) return weights;

  const learningRate = 0.05;

  for (const entry of decisions) {
    const rawFeatures = entry.features as Record<string, unknown>;
    const triageFeatures: TriageFeatures = {
      deadline: (rawFeatures.deadline as string) ?? null,
      amount: (rawFeatures.amount as string) ?? null,
      waiting_on_user: rawFeatures.waiting_on_user === true,
      category: String(rawFeatures.category ?? 'other'),
      urgency: (rawFeatures.urgency as 'high' | 'medium' | 'low') ?? 'low',
    };
    const fv = extractFeatureVector(triageFeatures);
    const direction = entry.decision === 'approve' ? 1 : -1;

    for (const key of FEATURE_KEYS) {
      if (fv[key] > 0) {
        weights[key] += learningRate * direction * fv[key];
      }
    }
  }

  // Clamp
  for (const key of FEATURE_KEYS) {
    weights[key] = Math.max(WEIGHT_BOUNDS[key].min, Math.min(WEIGHT_BOUNDS[key].max, weights[key]));
  }

  return weights;
}

export interface RankedCandidate {
  messageId: string;
  features: TriageFeatures;
  /** Lower = higher priority (position in queue). */
  score: number;
  /** Whether this candidate filled a floor-reserved slot. */
  floor: boolean;
  /** Whether this candidate fills an exploration slot. */
  exploration?: boolean;
  /** Per-feature score breakdown for debuggability. */
  breakdown?: ScoreBreakdown;
}

/**
 * Check if a candidate matches a floor reservation condition.
 */
export function matchesFloor(
  features: TriageFeatures,
  match: Record<string, unknown>,
  now: Date = new Date(),
): boolean {
  if (typeof match.deadline_within_hours === 'number' && features.deadline) {
    const deadline = new Date(features.deadline);
    const hoursUntil = (deadline.getTime() - now.getTime()) / (1000 * 60 * 60);
    return hoursUntil >= 0 && hoursUntil <= match.deadline_within_hours;
  }
  if (typeof match.category === 'string') {
    return features.category === match.category;
  }
  if (typeof match.urgency === 'string') {
    return features.urgency === match.urgency;
  }
  return false;
}

export interface RankInput {
  messageId: string;
  features: TriageFeatures;
}

export interface RankOptions {
  weights: FeatureWeights;
  explorationSlots: number;
  journalEntries: JournalEntry[];
}

/**
 * Rank candidates with deterministic floor reservations, learned feature scoring,
 * and exploration slots.
 *
 * Algorithm:
 * 1. For each floor reservation, find matching candidates and reserve up to `slots` for them.
 * 2. If exploration slots requested, reserve positions for high-uncertainty candidates.
 * 3. Fill remaining slots with candidates sorted by learned feature score.
 * 4. Cap at targetDepth.
 */
export function rankCandidates(
  candidates: RankInput[],
  floor: FloorReservation[],
  targetDepth: number,
  now: Date = new Date(),
  options?: RankOptions,
): RankedCandidate[] {
  const weights = options?.weights ?? DEFAULT_WEIGHTS;
  const explorationSlots = options?.explorationSlots ?? 0;
  const journalEntries = options?.journalEntries ?? [];

  const result: RankedCandidate[] = [];
  const used = new Set<string>();

  // Step 1: Fill floor-reserved slots
  for (const reservation of floor) {
    let filled = 0;
    for (const candidate of candidates) {
      if (used.has(candidate.messageId)) continue;
      if (filled >= reservation.slots) break;
      if (matchesFloor(candidate.features, reservation.match, now)) {
        const fv = extractFeatureVector(candidate.features, now);
        result.push({
          messageId: candidate.messageId,
          features: candidate.features,
          score: result.length,
          floor: true,
          breakdown: computeBreakdown(fv, weights),
        });
        used.add(candidate.messageId);
        filled++;
      }
    }
  }

  // Step 2: Reserve exploration slots for high-uncertainty candidates
  const explorationCandidates: RankedCandidate[] = [];
  if (explorationSlots > 0) {
    // Count how many times each messageId appears in journal
    const seenCounts = new Map<string, number>();
    for (const entry of journalEntries) {
      const mid = entry.messageId as string | undefined;
      if (mid) {
        seenCounts.set(mid, (seenCounts.get(mid) ?? 0) + 1);
      }
    }

    // Sort unused candidates by least-seen (highest uncertainty)
    const unseenCandidates = candidates
      .filter((c) => !used.has(c.messageId))
      .map((c) => ({ ...c, seenCount: seenCounts.get(c.messageId) ?? 0 }))
      .sort((a, b) => a.seenCount - b.seenCount);

    let filled = 0;
    for (const candidate of unseenCandidates) {
      if (filled >= explorationSlots) break;
      if (result.length >= targetDepth) break;
      const fv = extractFeatureVector(candidate.features, now);
      const ranked: RankedCandidate = {
        messageId: candidate.messageId,
        features: candidate.features,
        score: result.length,
        floor: false,
        exploration: true,
        breakdown: computeBreakdown(fv, weights),
      };
      explorationCandidates.push(ranked);
      used.add(candidate.messageId);
      filled++;
    }
  }

  // Step 3: Score and sort remaining candidates by learned feature score
  const scored = candidates
    .filter((c) => !used.has(c.messageId))
    .map((c) => {
      const fv = extractFeatureVector(c.features, now);
      return {
        ...c,
        fv,
        totalScore: scoreCandidate(fv, weights),
        breakdown: computeBreakdown(fv, weights),
      };
    })
    .sort((a, b) => b.totalScore - a.totalScore); // Higher score = higher priority

  for (const candidate of scored) {
    if (result.length + explorationCandidates.length >= targetDepth) break;
    result.push({
      messageId: candidate.messageId,
      features: candidate.features,
      score: result.length,
      floor: false,
      breakdown: candidate.breakdown,
    });
  }

  // Insert exploration candidates (at the end of the scored section)
  result.push(...explorationCandidates);

  // Cap at targetDepth (floor slots might have already exceeded it)
  return result.slice(0, targetDepth);
}

File: python/steward/ranker.py
"""Learned-over-features ranker with deterministic floor and exploration slots."""
from __future__ import annotations

from dataclasses import dataclass, field
from datetime import datetime, timezone
from typing import Any, TypedDict

from steward.rules import FloorReservation
from steward.triage import TriageFeatures


class FeatureVector(TypedDict):
    deadline_proximity: float
    has_amount: float
    waiting_on_user: float
    urgency: float


FEATURE_KEYS: tuple[str, ...] = (
    "deadline_proximity",
    "has_amount",
    "waiting_on_user",
    "urgency",
)

WEIGHT_BOUNDS: dict[str, tuple[float, float]] = {
    "deadline_proximity": (0.1, 3.0),
    "has_amount": (0.1, 3.0),
    "waiting_on_user": (0.1, 3.0),
    "urgency": (0.2, 3.0),
}

DEFAULT_WEIGHTS: dict[str, float] = {
    "deadline_proximity": 1.0,
    "has_amount": 0.5,
    "waiting_on_user": 0.8,
    "urgency": 1.0,
}


class ScoreBreakdown(TypedDict):
    deadline_proximity: float
    has_amount: float
    waiting_on_user: float
    urgency: float
    total: float


def _now() -> datetime:
    return datetime.now(timezone.utc)


def _parse_deadline(deadline: str | None) -> datetime | None:
    if not deadline:
        return None
    try:
        s = deadline.replace("Z", "+00:00")
        dt = datetime.fromisoformat(s)
        if dt.tzinfo is None:
            dt = dt.replace(tzinfo=timezone.utc)
        return dt
    except ValueError:
        return None


def extract_feature_vector(features: TriageFeatures, now: datetime | None = None) -> FeatureVector:
    n = now or _now()
    deadline_proximity = 0.0
    deadline_dt = _parse_deadline(features.get("deadline"))
    if deadline_dt is not None:
        hours_until = (deadline_dt - n).total_seconds() / 3600
        if 0 <= hours_until <= 168:
            deadline_proximity = max(0.0, 1.0 - hours_until / 168)

    urgency_map = {"high": 1.0, "medium": 0.5, "low": 0.0}
    return {
        "deadline_proximity": deadline_proximity,
        "has_amount": 1.0 if features.get("amount") is not None else 0.0,
        "waiting_on_user": 1.0 if features.get("waiting_on_user") else 0.0,
        "urgency": urgency_map.get(features.get("urgency", "low"), 0.0),
    }


def score_candidate(fv: FeatureVector, weights: dict[str, float]) -> float:
    return sum(fv[k] * weights[k] for k in FEATURE_KEYS)  # type: ignore[literal-required]


def compute_breakdown(fv: FeatureVector, weights: dict[str, float]) -> ScoreBreakdown:
    breakdown: dict[str, float] = {}
    total = 0.0
    for key in FEATURE_KEYS:
        c = fv[key] * weights[key]  # type: ignore[literal-required]
        breakdown[key] = c
        total += c
    breakdown["total"] = total
    return breakdown  # type: ignore[return-value]


def learn_weights(entries: list[dict[str, Any]]) -> dict[str, float]:
    weights = dict(DEFAULT_WEIGHTS)
    decisions = [
        e for e in entries
        if e.get("kind") == "decision"
        and e.get("decision") in ("approve", "reject", "defer")
        and e.get("features")
    ]
    if not decisions:
        return weights

    learning_rate = 0.05
    for entry in decisions:
        raw = entry["features"]
        tf: TriageFeatures = {
            "deadline": raw.get("deadline"),
            "amount": raw.get("amount"),
            "waiting_on_user": raw.get("waiting_on_user") is True,
            "category": raw.get("category") or "other",
            "urgency": raw.get("urgency") or "low",
        }
        fv = extract_feature_vector(tf)
        direction = 1 if entry["decision"] == "approve" else -1
        for key in FEATURE_KEYS:
            if fv[key] > 0:  # type: ignore[literal-required]
                weights[key] += learning_rate * direction * fv[key]  # type: ignore[literal-required]

    for key, (lo, hi) in WEIGHT_BOUNDS.items():
        weights[key] = max(lo, min(hi, weights[key]))
    return weights


@dataclass
class RankInput:
    messageId: str
    features: TriageFeatures


@dataclass
class RankedCandidate:
    messageId: str
    features: TriageFeatures
    score: int
    floor: bool
    exploration: bool = False
    breakdown: ScoreBreakdown | None = None


def matches_floor(
    features: TriageFeatures,
    match: dict[str, Any],
    now: datetime | None = None,
) -> bool:
    n = now or _now()
    if isinstance(match.get("deadline_within_hours"), (int, float)) and features.get("deadline"):
        deadline_dt = _parse_deadline(features["deadline"])
        if deadline_dt is None:
            return False
        hours_until = (deadline_dt - n).total_seconds() / 3600
        return 0 <= hours_until <= match["deadline_within_hours"]
    if isinstance(match.get("category"), str):
        return features.get("category") == match["category"]
    if isinstance(match.get("urgency"), str):
        return features.get("urgency") == match["urgency"]
    return False


@dataclass
class RankOptions:
    weights: dict[str, float] = field(default_factory=lambda: dict(DEFAULT_WEIGHTS))
    exploration_slots: int = 0
    journal_entries: list[dict[str, Any]] = field(default_factory=list)


def rank_candidates(
    candidates: list[RankInput],
    floor: list[FloorReservation],
    target_depth: int,
    now: datetime | None = None,
    options: RankOptions | None = None,
) -> list[RankedCandidate]:
    opts = options or RankOptions()
    weights = opts.weights
    exploration_slots = opts.exploration_slots
    journal_entries = opts.journal_entries
    n = now or _now()

    result: list[RankedCandidate] = []
    used: set[str] = set()

    # Step 1: Fill floor-reserved slots
    for reservation in floor:
        filled = 0
        for candidate in candidates:
            if candidate.messageId in used:
                continue
            if filled >= reservation.slots:
                break
            if matches_floor(candidate.features, reservation.match, n):
                fv = extract_feature_vector(candidate.features, n)
                result.append(
                    RankedCandidate(
                        messageId=candidate.messageId,
                        features=candidate.features,
                        score=len(result),
                        floor=True,
                        breakdown=compute_breakdown(fv, weights),
                    )
                )
                used.add(candidate.messageId)
                filled += 1

    # Step 2: Exploration slots
    exploration: list[RankedCandidate] = []
    if exploration_slots > 0:
        seen_counts: dict[str, int] = {}
        for entry in journal_entries:
            mid = entry.get("messageId")
            if isinstance(mid, str):
                seen_counts[mid] = seen_counts.get(mid, 0) + 1

        unseen = sorted(
            (c for c in candidates if c.messageId not in used),
            key=lambda c: seen_counts.get(c.messageId, 0),
        )
        filled = 0
        for candidate in unseen:
            if filled >= exploration_slots:
                break
            if len(result) >= target_depth:
                break
            fv = extract_feature_vector(candidate.features, n)
            exploration.append(
                RankedCandidate(
                    messageId=candidate.messageId,
                    features=candidate.features,
                    score=len(result),
                    floor=False,
                    exploration=True,
                    breakdown=compute_breakdown(fv, weights),
                )
            )
            used.add(candidate.messageId)
            filled += 1

    # Step 3: Score remaining
    remaining = [c for c in candidates if c.messageId not in used]
    scored = []
    for c in remaining:
        fv = extract_feature_vector(c.features, n)
        scored.append((c, fv, score_candidate(fv, weights), compute_breakdown(fv, weights)))
    scored.sort(key=lambda t: t[2], reverse=True)

    for c, fv, total, breakdown in scored:
        if len(result) + len(exploration) >= target_depth:
            break
        result.append(
            RankedCandidate(
                messageId=c.messageId,
                features=c.features,
                score=len(result),
                floor=False,
                breakdown=breakdown,
            )
        )

    result.extend(exploration)
    return result[:target_depth]
